File: apps/web/src/lib/literature/rate-limiter.ts
// Token bucket rate limiter for external API calls

interface RateLimiterConfig {
    maxTokens: number;
    refillRateMs: number; // ms between refills
}

class TokenBucketRateLimiter {
    private tokens: number;
    private lastRefill: number;
    private readonly maxTokens: number;
    private readonly refillRateMs: number;

    constructor(config: RateLimiterConfig) {
        this.maxTokens = config.maxTokens;
        this.refillRateMs = config.refillRateMs;
        this.tokens = config.maxTokens;
        this.lastRefill = Date.now();
    }

    private refill(): void {
        const now = Date.now();
        const elapsed = now - this.lastRefill;
        const tokensToAdd = Math.floor(elapsed / this.refillRateMs);

        if (tokensToAdd > 0) {
            this.tokens = Math.min(this.maxTokens, this.tokens + tokensToAdd);
            this.lastRefill = now;
        }
    }

    async acquire(): Promise<void> {
        this.refill();

        if (this.tokens > 0) {
            this.tokens--;
            return;
        }

        // Wait for next token
        const waitTime = this.refillRateMs - (Date.now() - this.lastRefill);
        await new Promise(resolve => setTimeout(resolve, Math.max(waitTime, 0)));
        this.refill();
        this.tokens--;
    }
}

// arXiv: 1 request per 3 seconds
export const arxivLimiter = new TokenBucketRateLimiter({
    maxTokens: 1,
    refillRateMs: 3000,
});

// Semantic Scholar: 10 requests per second (free tier, no API key)
export const semanticScholarLimiter = new TokenBucketRateLimiter({
    maxTokens: 10,
    refillRateMs: 1000,
});

// Papers With Code: ~5 requests per second (conservative)
export const pwcLimiter = new TokenBucketRateLimiter({
    maxTokens: 5,
    refillRateMs: 1000,
});

// GitHub: 60 req/hour unauthed (1/min), 5000 req/hour with token (~1.4/sec)
export const githubLimiter = new TokenBucketRateLimiter({
    maxTokens: process.env.GITHUB_TOKEN ? 5 : 1,
    refillRateMs: process.env.GITHUB_TOKEN ? 1000 : 60000,
});

File: services/hypothesis-room/src/vreda_hypothesis/utils/rate_limiter.py
"""Async token-bucket rate limiter for external API clients.

Prevents hitting API rate limits for arXiv (1 req/3s), Semantic Scholar (10 req/s),
and PapersWithCode (5 req/s).
"""

from __future__ import annotations

import asyncio
import time

import structlog

logger = structlog.get_logger(__name__)


class AsyncRateLimiter:
    """Token-bucket rate limiter for async API calls.

    Usage:
        limiter = AsyncRateLimiter(rate=10, period=1.0)  # 10 requests per second
        async with limiter:
            await make_api_call()
    """

    def __init__(self, rate: int, period: float = 1.0, name: str = "default") -> None:
        self._rate = rate
        self._period = period
        self._name = name
        self._tokens = float(rate)
        self._last_refill = time.monotonic()
        self._lock = asyncio.Lock()

    async def acquire(self) -> None:
        """Wait until a token is available."""
        while True:
            async with self._lock:
                now = time.monotonic()
                elapsed = now - self._last_refill
                self._tokens = min(
                    float(self._rate),
                    self._tokens + elapsed * (self._rate / self._period),
                )
                self._last_refill = now

                if self._tokens >= 1.0:
                    self._tokens -= 1.0
                    return

            # Wait for a token to become available
            wait_time = self._period / self._rate
            logger.debug("rate_limiter.waiting", name=self._name, wait=wait_time)
            await asyncio.sleep(wait_time)

    async def __aenter__(self) -> AsyncRateLimiter:
        await self.acquire()
        return self

    async def __aexit__(self, *args) -> None:
        pass


# Pre-configured limiters for external APIs
arxiv_limiter = AsyncRateLimiter(rate=1, period=3.0, name="arxiv")       # 1 req per 3 seconds
semantic_scholar_limiter = AsyncRateLimiter(rate=10, period=1.0, name="semantic_scholar")  # 10 req/s
paperswithcode_limiter = AsyncRateLimiter(rate=5, period=1.0, name="paperswithcode")       # 5 req/s
